# House Price Prediction using Machine Learning

## Phase 1: Dataset Understanding

### Objective
The objective of this phase is to understand the structure and contents of the Nepali House Price dataset before performing any preprocessing or model building. In this phase, we will inspect the dataset, identify the features, examine their data types, detect missing values, and understand the role of each attribute.

## Step 1: Import Required Libraries

We begin by importing the necessary Python libraries. Pandas is used for loading and analyzing the dataset, while NumPy provides support for numerical computations.

In [4]:
import numpy as np 
import pandas as pd

## Step 2: Load the Dataset

The dataset is loaded into a Pandas DataFrame using the `read_csv()` function. A DataFrame is a tabular data structure that makes data analysis and manipulation easier.

In [5]:
df=pd.read_csv("Nepali_house_dataset.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'Nepali_house_dataset.csv'

## Step 3: Preview First Few Rows

We use `head()` to display the first 5 rows of the dataset. This gives an initial look at the data — the column names, sample values, and general structure of each feature.

In [ ]:
print(df.head())

                     TITLE                LOCATION         PRICE LAND AREA  \
0           House for Sale        Imadol, Lalitpur   Rs. 2.9 Cr   4.0 aana   
1           House for Sale     Satdobato, Lalitpur  Rs. 4.75 Cr   3.0 aana   
2     4 BHK House for Sale        Imadol, Lalitpur  Rs. 1.99 Cr   2.3 aana   
3  Bungalow House for Sale    Bhaisepati, Lalitpur     Rs. 4 Cr   7.0 aana   
4           House for Rent  Maharajgunj, Kathmandu  Rs. 12000000  6.0 aana   

  BUILDUP AREA ROAD ACCESS      FACING  FLOOR  BEDROOM  BATHROOM BUILT YEAR  \
0          NaN     12 Feet        West    3.0      5.0       4.0   2076 B.S   
1          NaN     10 Feet        West    4.5      5.0       6.0   2076 B.S   
2          NaN     10 Feet        West    2.5      4.0       4.0   2060 B.S   
3          NaN     12 Feet  North-West    2.5      4.0       3.0   2059 B.S   
4          NaN     20 Feet       South    2.0      4.0       4.0   2071 B.S   

              PARKING                                   

## Step 4: Preview Last Few Rows

`tail()` shows the last 5 rows of the dataset. This is useful to verify that the data was loaded completely and to check whether the end of the file contains any anomalies or unexpected entries.

In [ ]:
print(df.tail())

                       TITLE                     LOCATION           PRICE  \
3413  Padma Colony Phase III        Sitapiala,  Kathmandu  Rs. 26,000,000   
3414   Bhatbhateni Apartment      Bhatbhateni,  Kathmandu  Rs. 30,000,000   
3415               स्यूचाटार        Syuchatar,  Kathmandu    Rs. 16000000   
3416   Sano Bharayang Colony   Sano Bharayang,  Kathmandu      Rs. 3.5 Cr   
3417  Buddhanilkatha Banglow   Budhanilkantha,  Kathmandu    Rs. 48000000   

      LAND AREA BUILDUP AREA ROAD ACCESS      FACING  FLOOR  BEDROOM  \
3413   4.5 aana          NaN     16 Feet       North    2.5      4.0   
3414   4.5 aana   1700 Sq.ft     16 Feet       South    3.0      3.0   
3415  4.7 aana    1700 Sq.ft     16 Feet       South    2.5      4.0   
3416   6.3 aana   3505 Sq.ft     16 Feet  North-West    2.5      5.0   
3417     8 aana       5 aana     16 Feet       North    2.0      5.0   

      BATHROOM BUILT YEAR            PARKING  \
3413       4.0   2076 B.S    1 Car & 3 Bikes   
3414    

## Step 5: Check Dataset Shape

`shape` returns a tuple `(rows, columns)` indicating the size of the dataset. Knowing the number of records and features upfront is essential for planning the preprocessing and modeling steps.

In [ ]:
print(df.shape)

(3418, 13)


## Step 6: List Column Names

`columns` returns all the feature names in the dataset. This helps us identify which attributes are available and verify that the column headers were read correctly.

In [ ]:
print(df.columns)

Index(['TITLE', 'LOCATION', 'PRICE', 'LAND AREA', 'BUILDUP AREA',
       'ROAD ACCESS', 'FACING', 'FLOOR', 'BEDROOM', 'BATHROOM', 'BUILT YEAR',
       'PARKING', 'AMENITIES'],
      dtype='str')


## Step 7: Dataset Info Overview

`df.info` provides a summary of the DataFrame, including column names, non-null counts, and data types. It is a quick way to spot missing values and verify whether each column has the expected data type.

In [ ]:
print(df.info)

<bound method DataFrame.info of                         TITLE                     LOCATION           PRICE  \
0              House for Sale             Imadol, Lalitpur     Rs. 2.9 Cr    
1              House for Sale          Satdobato, Lalitpur    Rs. 4.75 Cr    
2        4 BHK House for Sale             Imadol, Lalitpur    Rs. 1.99 Cr    
3     Bungalow House for Sale         Bhaisepati, Lalitpur       Rs. 4 Cr    
4              House for Rent       Maharajgunj, Kathmandu    Rs. 12000000   
...                       ...                          ...             ...   
3413   Padma Colony Phase III        Sitapiala,  Kathmandu  Rs. 26,000,000   
3414    Bhatbhateni Apartment      Bhatbhateni,  Kathmandu  Rs. 30,000,000   
3415                स्यूचाटार        Syuchatar,  Kathmandu    Rs. 16000000   
3416    Sano Bharayang Colony   Sano Bharayang,  Kathmandu      Rs. 3.5 Cr   
3417   Buddhanilkatha Banglow   Budhanilkantha,  Kathmandu    Rs. 48000000   

      LAND AREA BUILDUP AREA RO

## Step 8: Inspect Data Types

`dtypes` shows the data type of each column. Most columns here are stored as `str` (object), meaning they contain mixed or non-numeric text. Only `FLOOR`, `BEDROOM`, and `BATHROOM` are numeric (`float64`), which suggests significant cleaning will be needed for the remaining columns.

In [ ]:
print(df.dtypes)

TITLE               str
LOCATION            str
PRICE               str
LAND AREA           str
BUILDUP AREA        str
ROAD ACCESS         str
FACING              str
FLOOR           float64
BEDROOM         float64
BATHROOM        float64
BUILT YEAR          str
PARKING             str
AMENITIES           str
dtype: object


## Step 9: Count Missing Values

`isnull().sum()` counts the number of missing (NaN) values in each column. This is the first step in understanding how much data is absent and which columns will require imputation or removal during the cleaning phase.

In [ ]:
print(df.isnull().sum())

TITLE              0
LOCATION           0
PRICE              0
LAND AREA         89
BUILDUP AREA    2699
ROAD ACCESS        9
FACING           206
FLOOR             95
BEDROOM          282
BATHROOM         346
BUILT YEAR        61
PARKING         2786
AMENITIES          0
dtype: int64


## Step 10: Missing Value Percentage

Here we calculate the percentage of missing values per column. This gives a clearer picture of severity — for example, `PARKING` is missing ~81.5% of its values and `BUILDUP AREA` ~79%, making them candidates for either imputation or dropping during preprocessing.

In [ ]:
missing = ((df.isnull().sum() / len(df)) * 100).round(2)
print(missing)



TITLE            0.00
LOCATION         0.00
PRICE            0.00
LAND AREA        2.60
BUILDUP AREA    78.96
ROAD ACCESS      0.26
FACING           6.03
FLOOR            2.78
BEDROOM          8.25
BATHROOM        10.12
BUILT YEAR       1.78
PARKING         81.51
AMENITIES        0.00
dtype: float64


## Step 11: Check for Duplicate Rows

`duplicated().sum()` counts the number of fully duplicated rows. A result of `0` here confirms there are no exact duplicate entries in the dataset, so no rows need to be removed on this basis.

In [ ]:
print(df.duplicated().sum())

0


#### Step 12: Statistical Summary of Numerical Columns

`describe()` computes summary statistics (count, mean, std, min, quartiles, max) for numeric columns only — `FLOOR`, `BEDROOM`, and `BATHROOM`. Notably, the max value of 100 for both `BEDROOM` and `BATHROOM` suggests possible outliers that will need to be addressed.

In [ ]:
print(df.describe())

             FLOOR      BEDROOM     BATHROOM
count  3323.000000  3136.000000  3072.000000
mean      2.668492     5.399235     4.187826
std       0.726047     3.148436     2.674617
min       1.000000     1.000000     1.000000
25%       2.500000     4.000000     3.000000
50%       2.500000     5.000000     4.000000
75%       2.500000     6.000000     5.000000
max       8.000000   100.000000   100.000000


#### Step 13: Statistical Summary of All Columns

`describe(include='all')` extends the summary to cover all columns, including string/categorical ones. For non-numeric columns it shows `count`, `unique`, `top` (most frequent value), and `freq` (its frequency). This helps spot dominant categories and how varied each field is.

In [ ]:
print(df.describe(include='all'))

                  TITLE          LOCATION          PRICE LAND AREA  \
count              3418              3418           3418      3329   
unique             2895               514            537       210   
top      House for Sale  Imadol, Lalitpur  Price on call  4.0 aana   
freq                183               234            188       414   
mean                NaN               NaN            NaN       NaN   
std                 NaN               NaN            NaN       NaN   
min                 NaN               NaN            NaN       NaN   
25%                 NaN               NaN            NaN       NaN   
50%                 NaN               NaN            NaN       NaN   
75%                 NaN               NaN            NaN       NaN   
max                 NaN               NaN            NaN       NaN   

         BUILDUP AREA ROAD ACCESS FACING        FLOOR      BEDROOM  \
count             719        3409   3212  3323.000000  3136.000000   
unique            4

## Step 14: Unique Location Count

`nunique()` returns the number of distinct values in the `LOCATION` column. With 514 unique locations, this column is highly granular. During modeling, it may need to be grouped or encoded carefully to avoid overfitting.

In [ ]:
# Count the unique values
a=df['LOCATION'].nunique()
print(a)

514


## Step 15: Unique Value Count for All Columns

We loop through every column and print the count of unique values. This provides a quick cardinality check — high-cardinality columns like `TITLE` (2895) and `AMENITIES` (1713) will need special treatment, while low-cardinality ones like `FACING` (36) and `FLOOR` (15) are more manageable for encoding.

In [ ]:
for col in df.columns:
    print(col)
    print(df[col].nunique())
    print()
  


TITLE
2895

LOCATION
514

PRICE
537

LAND AREA
210

BUILDUP AREA
418

ROAD ACCESS
127

FACING
36

FLOOR
15

BEDROOM
30

BATHROOM
26

BUILT YEAR
39

PARKING
86

AMENITIES
1713



## Step 16: Sample Unique PRICE Values

We inspect a sample of unique price values. The output reveals inconsistent formatting — some entries use `Rs. X Cr` notation while others use raw numbers like `Rs. 12000000`. Standardizing this column into a single numeric format will be a key cleaning task.

In [ ]:

print(df["PRICE"].unique()[:10])

<StringArray>
[ 'Rs. 2.9 Cr ', 'Rs. 4.75 Cr ', 'Rs. 1.99 Cr ',    'Rs. 4 Cr ',
 'Rs. 12000000', 'Rs. 27000000',  'Rs. 3.3 Cr ',  'Rs. 4.5 Cr ',
     'Rs. 3 Cr', 'Rs. 4.99 Cr ']
Length: 10, dtype: str


## Step 17: Sample Unique LAND AREA Values

Inspecting unique values in `LAND AREA` shows a mix of units — `aana`, `kattha`, and `sq. mtr`. There are also malformed entries like `14.4.0 aana` and `0.0 aana`. All of these will need to be converted to a uniform unit during data cleaning.

In [ ]:
df["LAND AREA"].unique()[:50]

<StringArray>
[   '4.0 aana',    '3.0 aana',    '2.3 aana',    '7.0 aana',    '6.0 aana',
    '3.2 aana',    '4.3 aana',    '2.2 aana',    '5.0 aana',    '5.2 aana',
    '6.2 aana',    '4.2 aana',    '4.1 aana',    '9.6 aana',    '3.1 aana',
   '11.0 aana',    '3.3 aana',    '9.0 aana',     '12 aana',    '6.4 aana',
   '12.0 aana',    '3.5 aana',      '6 aana',    '7.1 aana',    '6.3 aana',
   '15.0 aana',   '12.1 aana',   '16.0 aana',    '9.3 aana',    '9.1 aana',
    '5.1 aana',   '28.0 aana', '0.14 kattha',   '24.0 aana', '14.4.0 aana',
   '13.2 aana',   '32.0 aana',    '0.0 aana',    '6.1 aana',    '2.9 aana',
    '7.2 aana',    '8.0 aana',   '10.2 aana',   '35.0 aana',   '37.0 aana',
   '10.0 aana', '127 sq. mtr',   '22.0 aana',   '30.0 aana',   '18.0 aana']
Length: 50, dtype: str

## Step 18: Sample Unique LOCATION Values

We sample 10 unique location values to understand how they are structured. Locations follow a `<Area>, <City>` pattern. Extracting the city (Kathmandu, Lalitpur, etc.) as a separate feature could be useful for modeling, as city-level pricing trends may be more meaningful than individual locality names.

In [ ]:
df["LOCATION"].unique()[:10]

<StringArray>
[         'Imadol, Lalitpur',       'Satdobato, Lalitpur',
      'Bhaisepati, Lalitpur',    'Maharajgunj, Kathmandu',
 'Budhanilkantha, Kathmandu',     'Gyaneshwor, Kathmandu',
       'Kharibot, Kathmandu',          'Bafal, Kathmandu',
       'Khumaltar, Lalitpur',       'Tahachal, Kathmandu']
Length: 10, dtype: str

## Step 19: Sample Unique ROAD ACCESS Values

`ROAD ACCESS` stores road width values like `12 Feet`, `16 Feet`, etc. The numeric part can be extracted and treated as a continuous feature. Wider road access is generally associated with higher property prices in Nepali real estate.

In [ ]:
df["ROAD ACCESS"].unique()[:10]



<StringArray>
['12 Feet', '10 Feet', '20 Feet', '13 Feet', '14 Feet', '16 Feet', '26 Feet',
 '25 Feet', '19 Feet', '18 Feet']
Length: 10, dtype: str

## Step 20: Sample Unique FACING Values

The `FACING` column contains the cardinal/intercardinal direction the property faces. There are inconsistencies — for example, both `South-East` and `South East` (with a space) appear as separate values. These will need to be normalized before encoding.

In [ ]:

df["FACING"].unique()[:10]


<StringArray>
[      'West', 'North-West',      'South', 'South-East', 'North-East',
       'East', 'South-West',      'North',          nan, 'South East']
Length: 10, dtype: str

## Step 21: Sample Unique PARKING Values

`PARKING` values are inconsistently formatted — `CaRs.` vs `Car`, extra spaces, and mixed patterns. Given the ~81.5% missing rate and messy format, this column will likely be dropped or heavily transformed. We may extract just the car count as a simpler numeric feature.

In [ ]:
df["PARKING"].unique()[:10]

<StringArray>
['1 CaRs.  & 2 Bikes', '2 CaRs.  & 2 Bikes', '1 CaRs.  & 3 Bikes',
 '4 CaRs.  & 4 Bikes', '4 CaRs.  & 5 Bikes',                  nan,
            '2 Bikes',            '3 Bikes',            '3 CaRs.',
 '1 CaRs.  & 1 Bikes']
Length: 10, dtype: str

## Step 22: Sample Unique AMENITIES Values

`AMENITIES` is stored as a string representation of a list. Each entry contains a varying set of features like `'Parking'`, `'Earthquake Resistant'`, `'Modular Kitchen'`, etc. To use this column in modeling, it will need to be parsed and converted into binary indicator columns (multi-label binarization) for each amenity.

In [ ]:
df["AMENITIES"].unique()[:5]

<StringArray>
[          '['Earthquake Resistant', 'Marbel', 'Parquet', 'Parking', 'Drinking Water', 'Terrace', 'Modular Kitchen', 'Closet', 'Dining Room', 'Pantry', 'Bed Room']',
      '['Earthquake Resistant', 'Parquet', 'Drinking Water', 'Parking', 'Modular Kitchen', 'Closet', 'Bed Room', 'Pantry', 'Dining Room', 'Bathroom', 'Store Room']',
        '['Earthquake Resistant', 'Marbel', 'Parquet', 'Parking', 'Drinking Water', 'Balcony', 'Closet', 'Modular Kitchen', 'MiCrowave', 'Bed Room', 'Dining Room']',
             '['Earthquake Resistant', 'Marbel', 'Parquet', 'Balcony', 'Parking', 'Drinking Water', 'Garden', 'Modular Kitchen', 'Internet', 'Closet', 'MiCrowave']',
 '['Earthquake Resistant', 'Parquet', 'Parking', 'Drinking Water', 'Closet', 'Modular Kitchen', 'Bed Room', 'Bathroom', 'Living Room', 'Guest Room', 'Prayer Room']']
Length: 5, dtype: str